src/components/cuda/cupti_profiler.c

```diff
@@ -1438,11 +1438,11 @@ int cuptip_control_create(cuptiu_event_table_t *event_names, cuptic_info_t thr_i
     if (papi_errno != PAPI_OK) {
         goto fn_exit;
     }
-    papi_errno = add_events_per_gpu(state, event_names);
+    papi_errno = nvpw_cuda_metricscontext_create(state);
     if (papi_errno != PAPI_OK) {
```

Comment:
How was this working at all? How can `add_events_per_gpu` take a state object that is created only after?

Reply:
`nvpw_cuda_metricscontext_create` creates CUPTI metrics contexts which are needed for enumeration of metrics, and not required by add events. Before, if add events failed, the function would exit without creating these, and on exit the component would try to destroy these and fail with segfault. So Creating these before add events successfully destroys these.

Reply:
I think naming is a little confusing. My interpretation of the code was that state is the control state object and `nvpw_cuda_metricscontext_create` creates the control state object, while `add_events_per_gpu` adds events to the control state object.

Reply:
`cuptip_control_create` creates the `state` object, then calls `nvpw_cuda_metricscontext_create` then `add_events_per_gpu`. `nvpw_cuda_metricscontext_create` creates the CUPTI MetricsContext object as members of the state object.